### PR TITLE
[FIX] mail: notify alias owner when fetching

### PR DESCRIPTION
**Pull request by reth-odoo**
[1] introduces an issue when a response is sent to an aliased email.

In that case, it is intended that the message should be created by the owner of the alias by [2]
Which breaks the assumption of made in the first commit, as the message is being passively processed by fetchmail.

This causes the owner of the alias to never be notified on reponses in cases where the message could be an alias update. i.e. on any model where the alias applies.
Because the responses are assumed to have been authored by the owner.

To avoid this issue while keeping the behaviors intact we check that we did not come from fetchmail.
If that is the case we consider the author is the stated author for notification purposes.

[1]: c676ed3ea906e99d27a6116cd77218c5ec95b416
[2]: af80c68ae50159d21977692c6d283c4067254edd

task-3383275

### CODE REVIEW

File: addons/test_mail/tests/test_mail_gateway.py [hunk 1]
```diff
@@ -751,27 +751,27 @@ def _employee_crash(*args, **kwargs):
             record = self.format_and_process(MAIL_TEMPLATE, self.user_employee.email_formatted, 'groups@test.com', subject='NoEmployeeAllowed')
         self.assertEqual(record.create_uid, self.user_employee)
         self.assertEqual(record.message_ids[0].subject, 'NoEmployeeAllowed')
-        self.assertEqual(record.message_ids[0].create_uid, self.user_employee)
+        self.assertEqual(record.message_ids[0].create_uid.id, SUPERUSER_ID, 'Message should be created by caller of message_process.')
```
Comment (tde-banana-odoo) [inline addons/test_mail/tests/test_mail_gateway.py:754]: Actually root is already assigned to ``self.user_root`` in mail common

File: addons/test_mail/tests/test_mail_gateway.py [hunk 2]
```diff
@@ -918,6 +918,69 @@ def test_message_route_alias_multiple_new(self):
...
+            'alias_model_id': Model.id,
+            'alias_name': 'reply_author_notify_test',
+            'alias_user_id': self.user_employee.id
+        })
+
+        test_record.message_subscribe((author_partner | self.user_employee.partner_id).ids)
+
+        messages = test_record.message_ids
+
+        odoobot = self.env.ref('base.user_root')
```
Comment (tde-banana-odoo) [inline addons/test_mail/tests/test_mail_gateway.py:950]: can use ``self.user_root``

File: addons/test_mail/tests/test_mail_gateway.py [hunk 3]
```diff
@@ -918,6 +918,69 @@ def test_message_route_alias_multiple_new(self):
...
+        params = ((email_to, user)
+                  for user in users
+                  for email_to in email_tos)
+        for email_to, user in params:
+            with self.subTest(user=user, email_to=email_to):
+                with self.mock_mail_gateway(), self.mock_mail_app():
+                    self.format_and_process(
+                        MAIL_TEMPLATE, self.email_from, email_to,
+                        subject=message.message_id, extra=f'In-Reply-To:\r\n\t{message.message_id}\n',
+                        model=None, user=user)
```
Comment (tde-banana-odoo) [inline addons/test_mail/tests/test_mail_gateway.py:964]: Not sure what user is doing here ? I think it ends up in formatting of email as available rendering variable, not sure it does more ?

File: addons/test_mail/tests/test_mail_gateway.py [hunk 4]
```diff
@@ -918,6 +918,69 @@ def test_message_route_alias_multiple_new(self):
...
+        test_record.message_subscribe((author_partner | self.user_employee.partner_id).ids)
+
+        messages = test_record.message_ids
+
+        odoobot = self.env.ref('base.user_root')
+        self.assertFalse(odoobot.active, 'notification logic relies on odoobot being archived')
+
+        users = [self.user_employee, odoobot]
+        email_tos = ['reply_author_notify_test@test.com', 'some_non_aliased_email@test.com']
+        params = ((email_to, user)
```
Comment (tde-banana-odoo) [inline addons/test_mail/tests/test_mail_gateway.py:955]: There is something like itertools.product that exists for that if I remember well

File: addons/mail/models/mail_thread.py [hunk 5]
```diff
@@ -1183,6 +1185,7 @@ def _message_route_process(self, message, message_dict, routes):
                 # postponed after message_post, because this is an external message and we don't want to create
                 # duplicate emails due to notifications
                 new_msg.write({'partner_ids': original_partner_ids})
+
```
Comment (tde-banana-odoo) [inline addons/mail/models/mail_thread.py:1188]: Unnecessary diff :disguised_face: (thanks tde)

**Review by tde-banana-odoo — COMMENTED** [review]
Zboing, thanks for the PR ! Added some remarks :) 

cc @Abridbus

File: addons/test_mail/tests/test_mail_gateway.py (hunk 3 above)
Comment (thiryr) [reply addons/test_mail/tests/test_mail_gateway.py:964]: Ooh right, I had a modification to format_and_process to have with_user and with_context in there and I removed when I removed the context :facepalm: 

I'll add it back, and use a different name

File: addons/mail/models/mail_thread.py [hunk 6]
```diff
@@ -1152,6 +1152,8 @@ def _message_route_process(self, message, message_dict, routes):
                 thread_id = thread.id
                 subtype_id = thread._creation_subtype().id
 
+            # switch to odoobot for all incoming message creation
+            thread = thread.with_user(self.env.ref('base.user_root'))
```
[the remark this answers is not in the document]
Comment (thiryr) [reply addons/mail/models/mail_thread.py:1156]: Yes, that's actually why I use odoobot instead of the user.

If we don't do this, calling message_process with some other user non-archived user will lead to the same issue. This is notably the case when:

- Clicking 'fetch now' in the incoming server settings
- Using a script that connects as some active user to call message_process to process .eml files directly (Our mail-gateway.py does this)

With this, we are assured the messages are posted with an inactive user and that our assumptions for authorship hold.

[the remark this answers is not in the document]
Comment (tde-banana-odoo) [reply addons/mail/models/mail_thread.py:1156]: Isn't it the role of the caller to maybe use sudo() before calling the message_process methods ?

File: addons/test_mail/tests/test_mail_gateway.py [hunk 7]
```diff
@@ -918,6 +919,65 @@ def test_message_route_alias_multiple_new(self):
...
+            'email': self.email_from,
+        })
+        message = self.env['mail.message'].create({
+            'body': '<p>test</p>',
+            'email_from': 'reply_author_notify_test@test.com',
+            'message_type': 'email',
+            'model': test_record._name,
+            'res_id': test_record.id,
+        })
+        Model = self.env['ir.model']._get(test_record._name)
```
Comment (tde-banana-odoo) [inline addons/test_mail/tests/test_mail_gateway.py:940]: can be inlined herebelow ?

File: addons/test_mail/tests/test_mail_gateway.py [hunk 8]
```diff
@@ -918,6 +919,65 @@ def test_message_route_alias_multiple_new(self):
...
+        """
+
+        test_record = self.env['mail.test.ticket'].create({})
+        author_partner = self.env['res.partner'].create({
+            'name': self.email_from,
+            'email': self.email_from,
+        })
+        message = self.env['mail.message'].create({
+            'body': '<p>test</p>',
+            'email_from': 'reply_author_notify_test@test.com',
```
Comment (tde-banana-odoo) [inline addons/test_mail/tests/test_mail_gateway.py:935]: Why has the message a from that is an alias ? This is quite not common normally ... is it related to the test ?

File: addons/test_mail/tests/test_mail_gateway.py [hunk 9]
```diff
@@ -918,6 +919,65 @@ def test_message_route_alias_multiple_new(self):
...
+        })
+
+        test_record.message_subscribe((author_partner | self.user_employee.partner_id).ids)
+
+        messages = test_record.message_ids
+
+        self.assertFalse(self.user_root.active, 'notification logic relies on odoobot being archived')
+
+        users = [self.user_employee, self.user_root]
+        email_tos = ['reply_author_notify_test@test.com', 'some_non_aliased_email@test.com']
```
Comment (tde-banana-odoo) [inline addons/test_mail/tests/test_mail_gateway.py:954]: Why two "tos", to test with and without alias usage ? Is it necessary for the test (aka: don't we already have tests checking a reply goes to the right thread even with a wrong email_to ?) ?

**Review by tde-banana-odoo — COMMENTED** [review]
Zboing zboing !

File: addons/test_mail/tests/test_mail_gateway.py (hunk 9 above)
Comment (thiryr) [reply addons/test_mail/tests/test_mail_gateway.py:954]: If the email_to is the alias email, we treat it differently than if it is some random email not linked to an alias, since we set the user to be the alias owner for update purposes.

The point is just to check that the behavior is consistent accross alias emails and non-alias emails for notifications on message replies.

File: addons/test_mail/tests/test_mail_gateway.py (hunk 8 above)
Comment (thiryr) [reply addons/test_mail/tests/test_mail_gateway.py:935]: email_from being the alias is relevant to cases where you set the alias to be your own email address, as a user. Such that e.g. whenever someone send an email to your user it creates a task for you to handle. 

In that case when you write a message it will be sent with your email as email_from, which is also the alias. Then when people respond to you, we detect they are responding to the alias and so we use the owner of the alias (also you) to write the message.

It's not necessary for the test but it's supposed to show why you would be responding to an alias in the first place, aside from people randomly deciding to edit the automatic email_to address when responding in their mail client.

----------------

Have added a comment to make that clearer but we can also remove it, it's really just to demonstrate the usecase.

File: addons/mail/models/mail_thread.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (thiryr) [reply addons/mail/models/mail_thread.py:1156]: With the current imp for 'real_author', which checks whether the user itself is active, this won't work because even if you sudo the current user doesn't actually change. You just set the su variable to bypass security checks.

'thread' is already sudoed before hand thanks to `ModelCtx = Model.with_user(related_user).sudo()` so switching to root entirely should be fine security-wise, or at least it will be as good as it currently is.

File: addons/test_mail/tests/test_mail_gateway.py (hunk 8 above)
Comment (thiryr) [reply addons/test_mail/tests/test_mail_gateway.py]: right, that makes more sense

I guess I can just remove it entirely then, yes

File: addons/mail/models/mail_alias.py [hunk 10]
```diff
@@ -42,7 +42,7 @@ class Alias(models.Model):
                                       # hack to only allow selecting mail_thread models (we might
                                       # (have a few false positives, though)
                                       domain="[('field_id.name', '=', 'message_ids')]")
-    alias_user_id = fields.Many2one('res.users', 'Owner', default=lambda self: self.env.user,
+    alias_user_id = fields.Many2one('res.users', 'Owner',
```
Comment (tde-banana-odoo) [inline addons/mail/models/mail_alias.py:45]: Note that first commit message still speaks about odoobot :)

File: addons/mail/models/mail_thread.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (tde-banana-odoo) [reply addons/mail/models/mail_thread.py:1156]: Note that we could also / or detect a su environment in the "real author" detection stuff :D . Would prefer seeing if we can do something with that, instead of the hard usage of "root" here that I don't really like.

**Review by tde-banana-odoo — COMMENTED** [review]
Zboing zboing !

File: addons/mail/models/mail_thread.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (thiryr) [reply addons/mail/models/mail_thread.py:1156]: Mm, I'm just worried that instances of 'sending a message as someone else' which the problematic feature addressed would also often involved sudo-ing which then goes against the point if we ignore the feature when using sudo

We can use context keys if you prefer but uh.. it would require adding a context key

[the remark this answers is not in the document]
Comment (tde-banana-odoo) [reply addons/mail/models/mail_thread.py]: All in all, ok with the usage of root then (just rename variable as discussed).

**Comment by tde-banana-odoo** [comment]
@robodoo r+ rebase-merge

**Comment by tde-banana-odoo** [comment]
@robodoo r+ rebase-merge
